src/librustc_trans/back/link.rs

```diff
@@ -259,13 +261,15 @@ pub fn filename_for_input(sess: &Session,
                           crate_name: &str,
                           outputs: &OutputFilenames) -> PathBuf {
     let libname = format!("{}{}", crate_name, sess.opts.cg.extra_filename);
+
+    if outputs.outputs.contains_key(&OutputType::Metadata) {
+        return outputs.out_directory.join(&format!("lib{}.rmeta", libname));
+    }
```

Comment:
This is where it may get a little tricky perhaps? It seems that the request for a metadata filename may want to be plumbed farther throughout the backend?